### PR TITLE
Add code actions doc

### PR DESCRIPTION
**Pull request by mindula**
Add VSCode document for code actions
Fixes: https://github.com/ballerina-platform/ballerina-lang/issues/39061

### CODE REVIEW

File: en/docs/edit-the-code/code-actions.md [hunk 1]
```diff
@@ -1 +1,70 @@
-# Code actions
\ No newline at end of file
+# Code actions
+
+## Types of Code actions
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:3]: We don't need this 2nd level heading. Instead let's just say
```
In VS Code, Code Actions can provide both refactorings and Quick Fixes for detected issues
```
(copied from vscode docs)

File: en/docs/edit-the-code/code-actions.md [hunk 2]
```diff
@@ -1 +1,70 @@
...
+
+### QuickFixes
+
+QuickFixes are suggested based on the diagnostic displayed at a given cursor position.
+
+#### Code Actions for Create Variables
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:9]: Shall we change this to `Quickfixes for variable assignment required` diagnostic. And move `Create variable with type` to a separate section like `Quickfixes for type inference`?

File: en/docs/edit-the-code/code-actions.md [hunk 3]
```diff
@@ -1 +1,70 @@
...
+
+The code action types below are available for creating a variable.
+- `Create variable`: Create a variable for an expression where the `Variable Assignment Required` diagnostic is present.
+- `Create variable and type guard`: Create a type guard to handle the error gracefully when the `Variable assignment Required` diagnostic is present.
+- `Create variable and check error`: Add a check expression when the `Variable assignment Required` diagnostic is present.
+- `Create variable with type`: Create a variable for a remote method expression where the `Cannot infer the 'typedesc' argument for parameter` diagnostic is present.
+- `Ignore return value`: Ignore the return value with the `_` where the `Variable Assignment Required` diagnostic is present.
+
+#### Code Actions for Union Types
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:18]:
```suggestion
#### Quick fixes for missing check expressions
```

File: en/docs/edit-the-code/code-actions.md [hunk 4]
```diff
@@ -1 +1,70 @@
...
+
+The code actions below are available for union-type variables.
+- `Add check error`: Add a check statement when there is an error union.
+
+#### Code Actions for Imports
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:23]: Let's name this section `Quick fixes for imports`. Same for the following topics under quick fixes section. Let's start the topic as `Quick fixes for xxx`

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: en/docs/edit-the-code/code-actions.md [hunk 5]
```diff
@@ -1 +1,149 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-incompatible-types.gif?raw=true" width="100%" />
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-incompatible-types2.gif?raw=true" width="100%" />
+
+#### Quick fixes for functions
+
+The code actions below are available for creating functions.
+- `Create function`: Create a function using the selected variables/parameters.
+- `Implement method`: Implement the selected method.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:77]:
```suggestion
- `Implement method`: Implement an unimplemented method declared by an included object type.
```

File: en/docs/edit-the-code/code-actions.md [hunk 6]
```diff
@@ -1 +1,149 @@
...
+- `Implement all`: Implement all the methods of an object.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:78]:
```suggestion
- `Implement all`: Implement all the methods declared by an included object type.
```

File: en/docs/edit-the-code/code-actions.md [hunk 7]
```diff
@@ -1 +1,149 @@
...
+- `Add explicit return statement`: Add an explicit return statement to a function.
+- `Convert to public function` : Add access modifier to a function.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:80]:
```suggestion
- `Convert to public function` : Add `public` access modifier to a function definition.
```

File: en/docs/edit-the-code/code-actions.md [hunk 8]
```diff
@@ -1 +1,149 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-functions.gif?raw=true" width="100%" />
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-functions2.gif?raw=true" width="100%" />
+
+#### Quick fixes for records
+
+The code actions below are available for records.
+- `Create record`: Create a record for an unknown type.
+- `Fill record fields`: Add missing record fields.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:91]:
```suggestion
- `Fill record fields`: Add missing record fields to a mapping constructor expression.
```

File: en/docs/edit-the-code/code-actions.md [hunk 9]
```diff
@@ -1 +1,149 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-records.gif?raw=true" width="100%" />
+
+#### Quick fixes for annotations
+
+The code actions below are available for annotations.
+- `Convert to constant`: Convert an annotation declaration to a constant declaration.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:98]:
```suggestion
- `Convert to constant`: Convert an annotation declaration to a constant annotation declaration.
```

File: en/docs/edit-the-code/code-actions.md [hunk 10]
```diff
@@ -1 +1,149 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-annotations.gif?raw=true" width="100%" />
+
+### Refactoring actions
+
+Refactoring actions are suggested when existing code can be rearranged or restructured to improve the quality and maintainability of your project while ensuring that the overall behavior of the code remains unchanged.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:104]:
```suggestion
Refactoring actions are suggested when existing code can be rearranged or restructured to improve the quality and maintainability of your code while ensuring that the overall behavior of the code remains unaffected.
```

File: en/docs/edit-the-code/code-actions.md [hunk 11]
```diff
@@ -1 +1,149 @@
...
+- `Extract to constant`: Extract a constant expression to a constant.
+- `Extract to function`: Extract an expression or a statement to a function.
+- `Extract to local variable`: Extract an expression to a local variable.
+- `Extract type`: Extract anonymous records to record types.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/extract-to-actions.gif?raw=true" width="100%" />
+
+#### Rewrite code actions
+
+- `Map with a query expression`: Convert RHS variable assignment to a query expression.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:117]:
```suggestion
- `Map with a query expression`: When both right hand side (RHS) and left hand side (LHS) types are array types, convert the RHS to a query expression. Users can later modify the query expression with additional clauses and mappings.
```

File: en/docs/edit-the-code/code-actions.md [hunk 12]
```diff
@@ -1 +1,149 @@
...
+
+- `Create a getter`: Generate a getter for class-level variables.
+- `Create a setter`: Generate a setter for class-level variables.
+- `Create getter and setter`: Generate both getter and setter for class-level variables.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/source-actions-for-getters-and-setters.gif?raw=true" width="100%" />
+
+#### Source actions for variables
+
+- `Type guard variable`: Type guard a variable if the variable is of the union type.
```
Comment (IMS94) [inline en/docs/edit-the-code/code-actions.md:135]:
```suggestion
- `Type guard variable`: Type guard a variable if the variable type is a union type.
```

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: en/docs/edit-the-code/code-actions.md [hunk 13]
```diff
@@ -1 +1,149 @@
...
+The code actions below are available for check expressions.
+- `Surround with do/on-fail`: Surround a check expression with do/on-fail clause. 
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-check-expressions.gif?raw=true" width="100%" />
+
+#### Quick fixes for imports
+
+The code actions below are available for imports.
+- `Import a module`: Add the import statement for a module, which has a reference without an import statement. This supports only the language library and the standard library.
+- `Pull unresolved module`: Pull locally unavailable Ballerina packages from the Ballerina Central repository.
```
[the remark this answers is not in the document]
Comment (mindula) [reply en/docs/edit-the-code/code-actions.md:44]: how about `Pull Ballerina package/packages from Ballerina central that is/are imported in the source and is/are unavailable locally.`?

File: en/docs/edit-the-code/code-actions.md [hunk 14]
```diff
@@ -1 +1,192 @@
...
+### QuickFixes
+
+QuickFixes are suggested based on the diagnostic displayed at a given cursor position.
+Ballerina VSCode plugin provides a wide range of quick fixes to such diagnostics.
+
+#### Quick fixes for variable assignment
+
+The code actions below are available for creating a variable.
+
+- `Create variable`: Create a variable for an expression where the `Variable assignment required` diagnostic is present.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:17]:
```suggestion
- `Create variable`: Create a variable for an expression in which the `Variable assignment required` diagnostic is present.
```

File: en/docs/edit-the-code/code-actions.md [hunk 15]
```diff
@@ -1 +1,192 @@
...
+- `Create variable and type guard`: Create a type guard to handle the error gracefully when the `Variable assignment required` diagnostic is present.
+- `Create variable and check error`: Add a check expression when the `Variable assignment required` diagnostic is present.
+- `Ignore return value`: Ignore the return value of a function call or a method call with `_` where the `Variable assignment required` diagnostic is present.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:20]:
```suggestion
- `Ignore return value`: Ignore the return value of a function call or a method call with `_` in which the `Variable assignment required` diagnostic is present.
```

File: en/docs/edit-the-code/code-actions.md [hunk 16]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-variable-assignments.gif?raw=true" width="100%" />
+
+#### Quick fixes for type inference
+
+The code actions below are available for type inference.
+
+- `Create variable with type`: Create a variable for a remote method expression where the `Cannot infer the 'typedesc' argument for parameter` diagnostic is present.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:28]:
```suggestion
- `Create variable with type`: Create a variable for a remote method expression in which the `Cannot infer the 'typedesc' argument for parameter` diagnostic is present.
```

File: en/docs/edit-the-code/code-actions.md [hunk 17]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-type-inference.gif?raw=true" width="100%" />
+
+#### Quick fixes for missing check expressions
+
+The code actions below are available for missing check expressions.
+
+- `Add 'check' error`: Add `check` keyword when the selected expression's type is a subtype of `error` or a union with `error` member while the expected type isn't/doesn't contain `error`.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:36]:
```suggestion
- `Add 'check' error`: Add the `check` keyword when the selected expression's type is a subtype of `error` or a union with an `error` member while the expected type isn't/doesn't contain an `error`.
```

File: en/docs/edit-the-code/code-actions.md [hunk 18]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-missing-check-expressions.gif?raw=true" width="100%" />
+
+#### Quick fixes for check expressions
+
+The code actions below are available for check expressions.
+
+- `Surround with do/on-fail`: Surround a check expression with do/on-fail clause.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:44]:
```suggestion
- `Surround with do/on-fail`: Surround a check expression with a do/on-fail clause.
```

File: en/docs/edit-the-code/code-actions.md [hunk 19]
```diff
@@ -1 +1,192 @@
...
+- `Change return type to 'error?'`: Change the return type of function/method to `error?`.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:45]:
```suggestion
- `Change return type to 'error?'`: Change the return type of a function/method to `error?`.
```

File: en/docs/edit-the-code/code-actions.md [hunk 20]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-documentation.gif?raw=true" width="100%" />
+
+#### Quick fixes for incompatible types
+
+The code actions below are available for incompatible types.
+
+- `Change variable type`: Change the type of variable.
+- `Add type cast`: Add a type cast for the incompatible type.
+- `Change return type`: Change the incompatible return type of function/method.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:75]:
```suggestion
- `Change return type`: Change the incompatible return type of a function/method.
```

File: en/docs/edit-the-code/code-actions.md [hunk 21]
```diff
@@ -1 +1,192 @@
...
+- `Change parameter type`: Change the type of function/ method parameter.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:76]:
```suggestion
- `Change parameter type`: Change the type of a function/ method parameter.
```

File: en/docs/edit-the-code/code-actions.md [hunk 22]
```diff
@@ -1 +1,192 @@
...
+- `Add conditional default value`: Add a conditional default value for the incompatible type.
+- `Convert to 'xml'/'string' template`: Convert a raw template into a subtype (`xml` or `string`).
+- `Add type cast for numeric expressions` Add a type cast for an unsupported numeric expression.
+
+`Change variable type`, `Change parameter type`, `Add type cast` and `Change return type` quick fixes in action.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:81]:
```suggestion
The `Change variable type`, `Change parameter type`, `Add type cast`, and `Change return type` quick fixes are demonstrated below.
```
This sentence is incomplete.

File: en/docs/edit-the-code/code-actions.md [hunk 23]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-incompatible-types.gif?raw=true" width="100%" />
+
+`Add conditional default value` and `Convert to 'xml'/'string' template` quick fixes in action.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:85]:
```suggestion
The `Add conditional default value` and `Convert to 'xml'/'string' template` quick fixes are demonstrated below.
```

File: en/docs/edit-the-code/code-actions.md [hunk 24]
```diff
@@ -1 +1,192 @@
...
+
+#### Quick fixes for functions
+
+The code actions below are available for creating functions.
+
+- `Create function`: Create a function when there's an undefined function invoked by a function call expression.
+- `Implement method`: Implement an unimplemented method declared by an included object type.
+- `Implement all`: Implement all the methods declared by an included object type.
+- `Add explicit return statement`: Add an explicit return statement to a function.
+- `Convert to public function` : Add `public` access modifier to a function definition.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:97]:
```suggestion
- `Convert to public function` : Add a `public` access modifier to a function definition.
```

File: en/docs/edit-the-code/code-actions.md [hunk 25]
```diff
@@ -1 +1,192 @@
...
+- `Add isolated qualifier`: Add isolated qualifier to a function.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:98]:
```suggestion
- `Add isolated qualifier`: Add an isolated qualifier to a function.
```

File: en/docs/edit-the-code/code-actions.md [hunk 26]
```diff
@@ -1 +1,192 @@
...
+
+`Create function`, `Implement method` and `Implement all` quick fixes in action.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:100]:
```suggestion
The `Create function`, `Implement method`, and `Implement all` quick fixes are demonstrated below.
```

File: en/docs/edit-the-code/code-actions.md [hunk 27]
```diff
@@ -1 +1,192 @@
...
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-functions.gif?raw=true" width="100%" />
+
+`Add explicit return statement`, `Convert to public function` and `Add isolated qualifier` quick fixes in action.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:104]:
```suggestion
The `Add explicit return statement`, `Convert to public function`, and `Add isolated qualifier` quick fixes are demonstrated below.
```

File: en/docs/edit-the-code/code-actions.md [hunk 28]
```diff
@@ -1 +1,192 @@
...
+
+The code actions below are available for annotations.
+
+- `Convert to constant`: Convert an annotation declaration to a constant annotation declaration.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-annotations.gif?raw=true" width="100%" />
+
+#### Other quick fixes
+
+- `Convert to readonly clone`: Convert an expression `e` to `e.cloneReadOnly()`.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:127]:
```suggestion
- `Convert to readonly clone`: Convert an `e` expression to `e.cloneReadOnly()`.
```

File: en/docs/edit-the-code/code-actions.md [hunk 29]
```diff
@@ -1 +1,192 @@
...
+- `Ignore unused variable`: Ignore an unused variable.
+- `Make construct pubic`: Make a construct public where the `attempt to expose non-public symbol` warning is present.
+- `Convert module variable to listener declaration`: Convert a module var declaration statement to a listener declaration.
+- `Remove unreachable code`: Remove unreachable statements.
+
+`Convert to readonly clone`, `Ignore unused variable` and `Remove unreachable code` quick fixes in action.
```
Comment (praneesha) [inline en/docs/edit-the-code/code-actions.md:133]:
```suggestion
The `Convert to readonly clone`, `Ignore unused variable` and `Remove unreachable code` quick fixes are demonstrated below. 
```

File: en/docs/edit-the-code/code-actions.md [hunk 30]
```diff
@@ -1 +1,192 @@
...
+
+The Ballerina VS Code extension provides code actions for the two scenarios below.
+
+1. Quick fixes for selected diagnostics
+2. Refactoring to improve code readability and suggest best practices
+
+### Quick fixes
+
+Quick fixes are suggested based on the diagnostics displayed at a given cursor position.
+The Ballerina VS Code extension provides a wide range of quick fixes to such diagnostics.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:11]: Is this the exclusive list of quick fixes for diagnostics? If so, we can say be more specific and say something like the following:

```
The Ballerina VS Code extension provides quick fixes for [diagnostics](link to page) identified in your Ballerina code as explained below.
```

@praneesha

File: en/docs/edit-the-code/code-actions.md [hunk 31]
```diff
@@ -1 +1,192 @@
...
+
+#### Quick fixes for variable assignment
+
+The code actions below are available for creating a variable.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:15]:
```suggestion
The following code actions are available as quick fixes when the `Variable assignment required` diagnostic is identified in your code:
```

File: en/docs/edit-the-code/code-actions.md [hunk 32]
```diff
@@ -1 +1,192 @@
...
+The `Convert module variable to listener declaration` quick fix is demonstrated below.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/other-quick-fixes3.gif?raw=true" width="100%" />
+
+### Refactoring actions
+
+Refactoring actions are suggested when the existing code can be rearranged or restructured to improve the quality and maintainability of your code while ensuring that the overall behavior of the code remains unaffected.
+
+#### Extract to code actions
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:150]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 33]
```diff
@@ -1 +1,192 @@
...
+#### Quick fixes for annotations
+
+The code actions below are available for annotations.
+
+- `Convert to constant`: Convert an annotation declaration to a constant annotation declaration.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-annotations.gif?raw=true" width="100%" />
+
+#### Other quick fixes
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:126]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 34]
```diff
@@ -1 +1,192 @@
...
+- `Extract to constant`: Extract a constant expression to a constant.
+- `Extract to function`: Extract an expression or a statement to a function.
+- `Extract to local variable`: Extract an expression to a local variable.
+- `Extract type`: Extract anonymous records to record types.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/extract-to-actions.gif?raw=true" width="100%" />
+
+#### Rewrite code actions
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:159]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 35]
```diff
@@ -1 +1,192 @@
-# Code actions
\ No newline at end of file
+# Code actions
+
+The Ballerina VS Code extension provides code actions for the two scenarios below.
+
+1. Quick fixes for selected diagnostics
+2. Refactoring to improve code readability and suggest best practices
+
+### Quick fixes
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:8]: ## and not ###

File: en/docs/edit-the-code/code-actions.md [hunk 36]
```diff
@@ -1 +1,192 @@
...
+- `Map with a query expression`: When both right-hand side (RHS) and left-hand side (LHS) types are array types, convert the RHS to a query expression. You can modify the query expression with additional clauses and mappings later.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/rewrite-actions.gif?raw=true" width="100%" />
+
+### Source actions
+
+Source actions are used to generate common code structures and recurring elements.
+
+#### Source actions for generating getters and setters
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:169]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 37]
```diff
@@ -1 +1,192 @@
...
+- `Create a getter`: Generate a getter for class-level variables.
+- `Create a setter`: Generate a setter for class-level variables.
+- `Create getter and setter`: Generate both getter and setter for class-level variables.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/source-actions-for-getters-and-setters.gif?raw=true" width="100%" />
+
+#### Source actions for variables
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:177]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 38]
```diff
@@ -1 +1,192 @@
...
+- `Type guard variable`: Type guard a variable if the variable type is a union type.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/source-actions-for-variable.gif?raw=true" width="100%" />
+
+#### Source actions for imports
+
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:183]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 39]
```diff
@@ -1 +1,192 @@
...
+- `Optimize imports`: Optimize import statements to remove unused imports and arrange imports in alphabetical order.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/source-actions-for-imports.gif?raw=true" width="100%" />
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:186]: intro

File: en/docs/edit-the-code/code-actions.md [hunk 40]
```diff
@@ -1 +1,192 @@
-# Code actions
\ No newline at end of file
+# Code actions
+
+The Ballerina VS Code extension provides code actions for the two scenarios below.
+
+1. Quick fixes for selected diagnostics
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:5]: What do we mean by `selected diagnostics`?

File: en/docs/edit-the-code/code-actions.md [hunk 41]
```diff
@@ -1 +1,192 @@
-# Code actions
\ No newline at end of file
+# Code actions
+
+The Ballerina VS Code extension provides code actions for the two scenarios below.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:3]: Is this a better intro:

```
[Code Actions](https://code.visualstudio.com/docs/editor/refactoring#_code-actions-quick-fixes-and-refactorings) is a feature in the VS Code Editor that enables quick fixes and code refactoring suggestions when issues are detected in the code. 

The Ballerina VS Code extension enables the following code actions when issues are detected in your Ballerina code.

> Note
   [Learn more](link to diagnostics page) about how to identify issues (diagnostics) in your Ballerina code.

```

File: en/docs/edit-the-code/code-actions.md (hunk 30 above)
Comment (mindula) [reply en/docs/edit-the-code/code-actions.md:11]: We do not provide quick fixes for every diagnostic in ballerina. Instead, for a small percentage of diagnostics, quick fixes are provided.

File: en/docs/edit-the-code/code-actions.md [hunk 42]
```diff
@@ -1 +1,202 @@
...
+
+- `Create variable with type`: Create a variable for a remote method expression in which the `Cannot infer the 'typedesc' argument for parameter` diagnostic is present.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-type-inference.gif?raw=true" width="100%" />
+
+#### Quick fixes for missing check expressions
+
+The following quick fixes are available for missing check expressions.
+
+- `Add 'check' error`: Add the `check` keyword when the selected expression's type is a subtype of `error` or a union with an `error` member while the expected type isn't/doesn't contain an `error`.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:34]:
```suggestion
- `Add 'check' error`: Add the `check` keyword when the selected expression's type is a subtype of `error` or a union with an `error` member while the expected type doesn't contain an `error`.
```

File: en/docs/edit-the-code/code-actions.md [hunk 43]
```diff
@@ -1 +1,202 @@
...
+- `Change return type to 'error?'`: Change the return type of a function/method to `error?`.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-check-expressions.gif?raw=true" width="100%" />
+
+#### Quick fixes for imports
+
+The following quick fixes are available for imports.
+
+- `Import module`: Add the import statement for a module that has a reference without an import statement.
+- `Pull unresolved module`: Pull Ballerina package/packages from Ballerina central that is/are imported in the source and is/are unavailable locally.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:52]:
```suggestion
- `Pull unresolved module`: Pull one or more Ballerina packages (which are imported in the source and are unavailable locally) from Ballerina central.
```

File: en/docs/edit-the-code/code-actions.md [hunk 44]
```diff
@@ -1 +1,202 @@
...
+- `Remove import`: Remove an unused or re-declared module import.
+- `Remove all unused imports`: Remove all unused imports except when there is a re-declared import statement.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/quick-fixes-for-imports.gif?raw=true" width="100%" />
+
+#### Quick fixes for documentation
+
+The following quick fixes are available for documentation.
+
+- `Document this`: Add the documentation to the closest documentable node.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:62]:
```suggestion
- `Document this`: Add the documentation to the closest node that can be documented.
```

File: en/docs/edit-the-code/code-actions.md [hunk 45]
```diff
@@ -1 +1,202 @@
...
+
+The following quick fixes are available for incompatible types.
+
+- `Change variable type`: Change the type of variable.
+- `Add type cast`: Add a type cast for the incompatible type.
+- `Change return type`: Change the incompatible return type of a function/method.
+- `Change parameter type`: Change the type of a function/method parameter.
+- `Add conditional default value`: Add a conditional default value for the incompatible type.
+- `Convert to 'xml'/'string' template`: Convert a raw template into a subtype (`xml` or `string`).
+- `Add type cast for numeric expressions` Add a type cast for an unsupported numeric expression.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:77]:
```suggestion
- `Add type cast for numeric expressions`: Add a type cast for an unsupported numeric expression.
```

File: en/docs/edit-the-code/code-actions.md [hunk 46]
```diff
@@ -1 +1,202 @@
...
+
+#### Other quick fixes
+
+- `Convert to readonly clone`: Convert an `e` expression to `e.cloneReadOnly()`.
+- `Ignore unused variable`: Ignore an unused variable.
+- `Make construct pubic`: Make a construct public where the `attempt to expose non-public symbol` warning is present.
+- `Convert module variable to listener declaration`: Convert a module var declaration statement to a listener declaration.
+- `Remove unreachable code`: Remove unreachable statements.
+
+The `Convert to readonly clone`, `Ignore unused variable` and `Remove unreachable code` quick fixes are demonstrated below.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:131]:
```suggestion
The `Convert to readonly clone`, `Ignore unused variable`, and `Remove unreachable code` quick fixes are demonstrated below.
```

File: en/docs/edit-the-code/code-actions.md [hunk 47]
```diff
@@ -1 +1,202 @@
...
+- `Extract to local variable`: Extract an expression to a local variable.
+- `Extract type`: Extract anonymous records to record types.
+
+  <img src="https://github.com/wso2/docs-ballerina-visual-studio-code-plugin/blob/main/en/docs/img/extract-to-actions.gif?raw=true" width="100%" />
+
+#### Rewrite code actions
+
+The following refactoring actions are available as rewrite actions.
+
+- `Map with a query expression`: When both right-hand side (RHS) and left-hand side (LHS) types are array types, convert the RHS to a query expression. You can modify the query expression with additional clauses and mappings later.
```
Comment (nilminiwso2) [inline en/docs/edit-the-code/code-actions.md:162]:
```suggestion
- `Map with a query expression`: When both the right-hand side (RHS) and left-hand side (LHS) types are array types, convert the RHS to a query expression. You can modify the query expression with additional clauses and mappings later.
```

**Review by nilminiwso2 — APPROVED** [review]

**Pull request merged**